pkg/salud/salud.go

```diff
@@ -0,0 +1,262 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package salud monitors the storage radius and request response duration of peers
+// and blocklists peers to maintain network salud (health).
+package salud
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/status"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+	"go.uber.org/atomic"
+)
+
+// loggerName is the tree path name of the logger for this package.
+const loggerName = "salud"
+
+const (
+	DefaultWakeup         = time.Minute
+	DefaultRequestTimeout = time.Second * 10
+)
+
+type topologyDriver interface {
+	UpdatePeerHealth(peer swarm.Address, health bool)
+	topology.PeerIterator
+}
+
+type peerStatus interface {
+	PeerSnapshot(ctx context.Context, peer swarm.Address) (*status.Snapshot, error)
+}
+
+type service struct {
+	wg            sync.WaitGroup
+	quit          chan struct{}
+	logger        log.Logger
+	topology      topologyDriver
+	status        peerStatus
+	metrics       metrics
+	isSelfHealthy *atomic.Bool
+	rs            postage.Radius
+}
+
+func New(status peerStatus, topology topologyDriver, rs postage.Radius, logger log.Logger, warmup time.Duration) *service {
+
+	metrics := newMetrics()
+
+	s := &service{
+		quit:          make(chan struct{}),
+		logger:        logger.WithName(loggerName).Register(),
+		status:        status,
+		topology:      topology,
+		metrics:       metrics,
+		isSelfHealthy: atomic.NewBool(true),
+		rs:            rs,
+	}
+
+	s.wg.Add(1)
+	go s.worker(warmup)
+
+	return s
+
+}
+
+func (s *service) worker(warmup time.Duration) {
+	defer s.wg.Done()
+
+	select {
+	case <-s.quit:
+		return
+	case <-time.After(warmup):
+	}
+
+	for {
+
+		s.salud()
+
+		select {
+		case <-s.quit:
+			return
+		case <-time.After(DefaultWakeup):
+		}
+	}
+}
+
+func (s *service) Close() error {
+	close(s.quit)
+	s.wg.Wait()
+	return nil
+}
+
+type peer struct {
+	status *status.Snapshot
+	dur    float64
+	addr   swarm.Address
+}
+
+// salud acquires the status snapshot of every peer and computes an avg response duration
+// and the most common storage radius and based on these values, it blocklist peers that fall beyond
+// some allowed threshold.
+func (s *service) salud() {
+
+	var (
+		mtx       sync.Mutex
+		wg        sync.WaitGroup
+		totaldur  float64
+		peers     []peer
+		neighbors int
+	)
+
+	_ = s.topology.EachConnectedPeer(func(addr swarm.Address, _ uint8) (stop bool, jumpToNext bool, err error) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), DefaultRequestTimeout)
+			defer cancel()
+
+			start := time.Now()
+
+			snapshot, err := s.status.PeerSnapshot(ctx, addr)
+			if err != nil {
+				s.topology.UpdatePeerHealth(addr, false)
+				return
+			}
+
+			if snapshot.BeeMode != "full" {
```

Comment:
This should be compared to `api.FullMode`. If it creates circular dependency, I'd suggest factoring out `BeeNodeMode` type from the `api` package.

Reply:
where should we put it?